### PR TITLE
Feature/accordion rewrite

### LINKED ISSUE #2312: Accordion: Automatically resize the Accordion if the content changes.
In SPAs (Single Page Application). The content of the 'Accordion' might dynamically change based on user input or async data fetching.

At the moment 'Accordion' does not dynamically resize if the content changes. So if the height of the content increases the 'Accordion' does not show all the content.

Possible use-cases:
- async loading / rendering of data after the Accordion is opended
- content added to Accordion based on user actions
- Accordion inside Accordion (mind-blowing ;-) )

Proposed implementation: Use an implementation based on ResizeObserver.

We have implemented a workaround for this in https://github.axa.com/avengers/bra-frontend (AccordionDynamic) that is dependent on implementation details of 'Accordion'.

Best regards

### PR DESCRIPTION
**Pull request by MKaHead**
fix #2312

It's a rewrite with more HTML semantic and remove all unessary DOM Manipulations.

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch-webhub-cloud/pattern-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.
- [x] Dependencies to other components still work.

### CODE REVIEW

File: src/components/20-molecules/accordion/index.js [hunk 1]
```diff
@@ -5,11 +5,51 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+import applyDefaults from '../../../utils/apply-defaults';
 import fireCustomEvent from '../../../utils/custom-event';
+import createRefId from '../../../utils/create-ref-id';
 import styles from './index.scss';
 
+const collapseContent = el => {
+  // get the height of the element's inner content, regardless of its actual size
+  const contentHeight = el.scrollHeight;
+  // temporarily disable all css transitions
+  const elementTransition = el.style.transition;
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: You are **saving** the existing transition here, to be used in a closure under rAF below. So, name it accordingly, e.g. savedElementTransition or savedTransition.

File: src/components/20-molecules/accordion/index.js [hunk 2]
```diff
@@ -5,11 +5,51 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+  el.style.transition = '';
+
+  // on the next frame (as soon as the previous style change has taken effect),
+  // explicitly set the element's height to its current pixel height, so we
+  // aren't transitioning out of 'auto'
+  requestAnimationFrame(() => {
+    el.style.height = `${contentHeight}px`;
+    el.style.transition = elementTransition;
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: see above.

File: src/components/20-molecules/accordion/index.js [hunk 3]
```diff
@@ -5,11 +5,51 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+
+    // on the next frame (as soon as the previous style change has taken effect),
+    // have the element transition to height: 0
+    requestAnimationFrame(() => {
+      el.style.height = '0px';
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: Make that '0', more concise.

File: src/components/20-molecules/accordion/index.js [hunk 4]
```diff
@@ -5,11 +5,51 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+
+  const expandAnimation = () => {
+    // remove "height" from the element's inline styles, so it can return to its initial value
+    el.style.height = null;
+  };
+
+  // when the next css transition finishes (which should be the one we just triggered)
+  el.addEventListener('transitionend', expandAnimation);
+  // remove this event listener so it only gets triggered once
+  el.removeEventListener('transitionend', expandAnimation);
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: Remove that in favour of option {once: true} in l.48 (https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#parameters)

File: src/components/20-molecules/accordion/index.js [hunk 5]
```diff
@@ -35,100 +77,83 @@ class AXAAccordion extends LitElement {
...
-            <slot></slot>
-          </div>
+          </button>
         </div>
-      </article>
+        <div
+          id="${this.refId}-content"
+          role="region"
+          aria-labelledby="${this.refId}"
+          class="m-accordion__content ${open
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: missed opportunity to use classMap, which would make this less brittle- please adapt

Comment (markus-walther) [reply src/components/20-molecules/accordion/index.js]: General rule for me would be: once you use classMap in a component, use it  consistently wherever a class list is not completely static!

File: src/components/20-molecules/accordion/react/test/dynamic/e2e.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+test.describe('button react: dynamic', () => {
+  test('should change text on state change', async ({ page }) => {
+    await page.goto(
+      fixtureURL('examples-accordion-react--demo-accordion-on-state-change')
+    );
+
+    await page.click('text=Test Accordion on state change');
+
+    expect(await page.textContent('text=true')).toBe('true');
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/react/test/dynamic/e2e.js:12]: Tests look really nice and clean now, bravo.

File: src/components/20-molecules/accordion/test/basic/e2e.js [hunk 7]
```diff
@@ -53,50 +53,8 @@ test.describe('accordion', () => {
 
     expect(
       await page
-        .locator('.m-accordion__title-container--disabled >> nth=0')
+        .locator('.m-accordion__button >> nth=0')
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/test/basic/e2e.js:57]: Had to find https://playwright.dev/docs/selectors#n-th-element-selector to understand this, might be worth a comment pointing there? Not sure.

File: src/components/20-molecules/accordion/index.js [hunk 8]
```diff
@@ -23,7 +63,9 @@ class AXAAccordion extends LitElement {
 
   static get properties() {
     return {
+      refId: { type: String, defaultValue: `accordion-${createRefId()}` },
       small: { type: Boolean },
+      level: { type: String },
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: Since it's a new property, we have the chance to make it crystal-clear: would ariaLevel be better?

**Review by markus-walther — CHANGES_REQUESTED** [review]
Hi Marco, 

good work! Few findings, plus: README and CHANGELOG need updates, too, right?

Cheers, Markus

File: src/components/20-molecules/accordion/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,9 @@
+## 1.2.0
+
+- Add the `ariaLevel` attribute defines the hierarchical level of an element within a structure.
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/CHANGELOG.md]: grammar: ... **which** defines the ....

File: src/components/20-molecules/accordion/CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,9 @@
...
+- Add more HTML semantic and remove all unnecessary DOM Manipulations
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/CHANGELOG.md]: spelling: **semantics** (noun, not adjective), **manipulation** (lowercase, singular)

File: src/components/20-molecules/accordion/index.js [hunk 11]
```diff
@@ -5,11 +5,49 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+const collapseContent = el => {
+  // get the height of the element's inner content, regardless of its actual size
+  const contentHeight = el.scrollHeight;
+  // temporarily disable all css transitions
+  const savedTransition = el.style.transition;
+  el.style.transition = '';
+
+  // on the next frame (as soon as the previous style change has taken effect),
+  // explicitly set the element's height to its current pixel height, so we
+  // aren't transitioning out of 'auto'
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: A bit clearer would be: aren't transitioning out of '**height:** auto'.

File: src/components/20-molecules/accordion/index.js [hunk 12]
```diff
@@ -5,11 +5,49 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+    requestAnimationFrame(() => {
+      el.style.height = '0';
+    });
+  });
+};
+
+const expandContent = el => {
+  // get the height of the element's inner content, regardless of its actual size
+  const contentHeight = el.scrollHeight;
+  // have the element transition to the height of its inner content
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js:39]: nice comments, to the point!

File: src/components/20-molecules/accordion/index.js [hunk 13]
```diff
@@ -5,11 +5,49 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+  el.style.height = `${contentHeight}px`;
+
+  const expandAnimation = () => {
+    // remove "height" from the element's inline styles, so it can return to its initial value
+    el.style.height = null;
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: question: when I read remove, I expected a `delete el.style.height;`. Would that not work here, and if not, why?

File: src/components/20-molecules/accordion/index.js [hunk 14]
```diff
@@ -5,11 +5,49 @@ import expandMore from '@axa-ch/materials/icons/material-design/expand_more.svg'
...
+  };
+
+  // when the next css transition finishes (which should be the one we just triggered)
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js]: spelling: **CSS** (all caps)

File: src/components/20-molecules/accordion/index.js [hunk 15]
```diff
@@ -23,7 +61,9 @@ class AXAAccordion extends LitElement {
 
   static get properties() {
     return {
+      refId: { type: String, defaultValue: `accordion-${createRefId()}` },
       small: { type: Boolean },
+      ariaLevel: { type: String },
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js:66]: yeah, now it is self-explanatory. Nice!

File: src/components/20-molecules/accordion/index.js [hunk 16]
```diff
@@ -35,100 +75,85 @@ class AXAAccordion extends LitElement {
   constructor() {
     super();
     applyDefaults(this);
-    this.toggleAnimation = this.toggleAnimation.bind(this);
+    this.toggleAccordion = this.toggleAccordion.bind(this);
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.js:78]: ah, better name!

File: src/components/20-molecules/accordion/index.scss [hunk 17]
```diff
@@ -1,96 +1,90 @@
...
-      margin: -2px;
-      outline: none;
-    }
-
+    &:focus,
     &:active {
       background-color: rgba($color-axa-blue, 0.2);
-      border: 2px solid $color-axa-blue;
-      margin: -2px;
+      box-shadow: 0 0 0 2px $color-axa-blue;
```
Comment (markus-walther) [inline src/components/20-molecules/accordion/index.scss:67]: Interesting. I guess it's more subtle this way?

**Review by markus-walther — CHANGES_REQUESTED** [review]
Hi Marco,

great work, just small findings mostly in the docs and comments.

Should be finished in no time.

Cheers, Markus

File: src/components/20-molecules/accordion/index.scss (hunk 17 above)
Comment (MKaHead) [reply src/components/20-molecules/accordion/index.scss:67]: yes, no problems with box sizing

**Review by markus-walther — APPROVED** [review]

**Pull request merged**